ports/esp32s2/common-hal/alarm/touch/TouchAlarm.c

```diff
@@ -0,0 +1,98 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "shared-bindings/alarm/touch/TouchAlarm.h"
+#include "shared-bindings/microcontroller/__init__.h"
+
+#include "peripherals/touch.h"
+
+#include "esp_sleep.h"
+
+void common_hal_alarm_touch_touchalarm_construct(alarm_touch_touchalarm_obj_t *self, const mcu_pin_obj_t *pin) {
+    if (pin->touch_channel == TOUCH_PAD_MAX) {
+        mp_raise_ValueError(translate("Invalid pin"));
+    }
+    claim_pin(pin);
+    self->pin = pin;
+}
+
+mp_obj_t alarm_touch_touchalarm_get_wakeup_alarm(size_t n_alarms, const mp_obj_t *alarms) {
+    // First, check to see if we match any given alarms.
+    for (size_t i = 0; i < n_alarms; i++) {
+        if (MP_OBJ_IS_TYPE(alarms[i], &alarm_touch_touchalarm_type)) {
+            return alarms[i];
+        }
+    }
+
+    gpio_num_t pin_number = esp_sleep_get_touchpad_wakeup_status();
+
+    alarm_touch_touchalarm_obj_t *alarm = m_new_obj(alarm_touch_touchalarm_obj_t);
+    alarm->base.type = &alarm_touch_touchalarm_type;
+    alarm->pin = NULL;
+
+    // Map the pin number back to a pin object.
+    for (size_t i = 0; i < mcu_pin_globals.map.used; i++) {
+        const mcu_pin_obj_t* pin_obj = MP_OBJ_TO_PTR(mcu_pin_globals.map.table[i].value);
+        if (pin_obj->number == pin_number) {
+            alarm->pin = mcu_pin_globals.map.table[i].value;
+            break;
+        }
+    }
+
+    return alarm;
+}
+
+static touch_pad_t touch_channel = TOUCH_PAD_MAX;
+
+void alarm_touch_touchalarm_set_alarm(alarm_touch_touchalarm_obj_t *self) {
+    touch_channel = (touch_pad_t)self->pin->number;
+    esp_sleep_enable_touchpad_wakeup();
+    esp_sleep_pd_config(ESP_PD_DOMAIN_RTC_PERIPH, ESP_PD_OPTION_ON);
+}
+
+void alarm_touch_touchalarm_prepare_for_deep_sleep(void) {
+    // intialize touchpad
+    peripherals_touch_init(touch_channel);
+
+    // configure touchpad for sleep
+    touch_pad_sleep_channel_enable(touch_channel, true);
+    touch_pad_sleep_channel_enable_proximity(touch_channel, false);
+
+    // wait for touch data to reset
+    mp_hal_delay_ms(10);
+
+    uint32_t touch_value;
+    touch_pad_sleep_channel_read_smooth(touch_channel, &touch_value);
+    touch_pad_sleep_set_threshold(touch_channel, touch_value * 0.1); //10%
+}
+
+bool alarm_touch_touchalarm_woke_us_up(void) {
+    return false;
```

Comment:
Does this work when pretending to deep sleep? Generally, you can use an internal callback to notify the main CP task to wake back up.

Reply:
Now using touch interrupt for `pretend-to-sleep`.

Reply:
I don't see any new changes. Did you forget to push them?